File: VulkanGameEngine/Sprite.cpp
#include "Sprite.h"
#include "GameObjectSystem.h"
#include "RenderSystem.h"
#include "LevelSystem.h"
#include "MaterialSystem.h"

uint32 Sprite::NextSpriteID = 0;

Sprite::Sprite()
{
}

Sprite::Sprite(GameObjectID gameObjectId, VkGuid& spriteVramId)
{
	SpriteID = ++NextSpriteID;
    GameObjectId = gameObjectId;
    SpriteVramId = spriteVramId;
	CurrentAnimationID = kStanding;
}

Sprite::~Sprite()
{
}

SpriteInstanceStruct Sprite::Update(VkCommandBuffer& commandBuffer, const float& deltaTime)
{
    const Transform2DComponent& transform2D = gameObjectSystem.FindTransform2DComponent(GameObjectId);
    const SpriteVram& vram = levelSystem.VramSpriteMap.at(SpriteVramId);
    const Animation2D& animation = levelSystem.AnimationMap.at(CurrentAnimationID);
    const AnimationFrames& frameList = levelSystem.AnimationFrameListMap[vram.VramSpriteID][CurrentAnimationID];
    const Material& material = materialSystem.FindMaterial(vram.SpriteMaterialID);
    const ivec2& currentFrame = frameList[CurrentFrame];

    mat4 spriteMatrix = mat4(1.0f);
    if (LastSpritePosition != SpritePosition)
    {
        spriteMatrix = glm::translate(spriteMatrix, vec3(transform2D.GameObjectPosition.x, transform2D.GameObjectPosition.y, 0.0f));
        LastSpritePosition == SpritePosition;
    }
    if (LastSpriteRotation != SpriteRotation)
    {
        spriteMatrix = glm::rotate(spriteMatrix, glm::radians(transform2D.GameObjectRotation.x), vec3(1.0f, 0.0f, 0.0f));
        spriteMatrix = glm::rotate(spriteMatrix, glm::radians(transform2D.GameObjectRotation.y), vec3(0.0f, 1.0f, 0.0f));
        spriteMatrix = glm::rotate(spriteMatrix, glm::radians(0.0f), vec3(0.0f, 0.0f, 1.0f));
        LastSpriteRotation == SpriteRotation;
    }
    if (LastSpriteScale != SpriteScale)
    {
        spriteMatrix = glm::scale(spriteMatrix, vec3(transform2D.GameObjectScale.x, transform2D.GameObjectScale.y, 1.0f));
        LastSpriteScale == SpriteScale;
    }

    CurrentFrameTime += deltaTime;
    if (CurrentFrameTime >= animation.FrameHoldTime) {
        CurrentFrame += 1;
        CurrentFrameTime = 0.0f;
        if (CurrentFrame >= frameList.size()) 
        {
            CurrentFrame = 0;
        }
    }

    SpriteInstanceStruct spriteInstance;
    spriteInstance.SpritePosition = transform2D.GameObjectPosition;
    spriteInstance.SpriteSize = vram.SpriteSize;
    spriteInstance.MaterialID = material.ShaderMaterialBufferIndex;
    spriteInstance.InstanceTransform = spriteMatrix;
    spriteInstance.FlipSprite = FlipSprite;
    spriteInstance.UVOffset = vec4(vram.SpriteUVSize.x * currentFrame.x, vram.SpriteUVSize.y * currentFrame.y, vram.SpriteUVSize.x, vram.SpriteUVSize.y);

    return spriteInstance;
}

void Sprite::SetSpriteAnimation(SpriteAnimationEnum spriteAnimation)
{
    if (CurrentAnimationID == spriteAnimation)
    {
        return;
    }

    CurrentAnimationID = spriteAnimation;
    CurrentFrame = 0;
    CurrentFrameTime = 0.0f;
}

File: VulkanGameEngineLevelEditor/GameEngine/Structs/Sprite.cs
using GlmSharp;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace VulkanGameEngineLevelEditor.GameEngine.Structs
{
    public struct Sprite
    {
        public enum SpriteAnimationEnum
        {
            kStanding,
            kWalking
        };

        [IgnoreProperty]
        public uint GameObjectId { get; set; }
        [ReadOnly(true)]
        public uint SpriteID { get; set; } = 0;
        [IgnoreProperty]
        public uint CurrentAnimationID { get; set; } = 0;
        [IgnoreProperty]
        public uint CurrentFrame { get; set; } = 0;
        public Guid SpriteVramId { get; set; }
        [IgnoreProperty]
        public float CurrentFrameTime { get; set; } = 0.0f;
        [IgnoreProperty]
        public bool SpriteAlive { get; set; } = true;
        [IgnoreProperty]
        public ivec2 FlipSprite { get; set; } = new ivec2(0);
        [IgnoreProperty]
        public vec2 LastSpritePosition { get; set; } = new vec2(0.0f);
        [IgnoreProperty]
        public vec2 LastSpriteRotation { get; set; } = new vec2(0.0f);
        [IgnoreProperty]
        public vec2 LastSpriteScale { get; set; } = new vec2(1.0f);
        public vec2 SpritePosition { get; set; } = new vec2(0.0f);
        public vec2 SpriteRotation { get; set; } = new vec2(0.0f);
        public vec2 SpriteScale { get; set; } = new vec2(1.0f);

        public Sprite()
        {
        }
    };

}
